### DISCUSSION TITLE
Why `.full()` and `.json()` are not present for `tracing_subscriber::fmt()` when `.pretty()` and `.compact()` are?

### DISCUSSION THREAD

**Discussion by frederikhors** [tokio-rs/tracing#2264] [Q&A] [answered]
I'm using this code instead of using `registry`:

```rust
tracing_subscriber::fmt()
  // .full() // this method is not present
  // .compact()
  // .pretty()
  // .json() // this method is not present
  .init();
```

Why those methods are not present?

**Comment by davidbarsky** [comment, marked as answer]
While `.full()` is already the default state of the builder returned by `tracing_subscriber::fmt()`, we should probably have that method for consistency's sake. `.json()` [_is_ present](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.SubscriberBuilder.html#method.json): did you enable the JSON feature in your `Cargo.toml`?

**Reply by frederikhors** [reply]
Oh. Wow. I'll check now.

A quick question: it's fine to use `tracing_subscriber::fmt().init()` directly or should I use `registry` with `tracing_subscriber::fmt::layer()`?

**Reply by frederikhors** [reply]
It works amazingly well enabling the JSON feature in my Cargo.toml.

Another quick question: how can I found on the docs pages that I need JSON feature reading [this page](https://docs.rs/tracing-subscriber/0.3.15/tracing_subscriber/fmt/format/index.html) and [this other page](https://docs.rs/tracing-subscriber/0.3.15/tracing_subscriber/fmt/format/struct.Json.html)?

**Reply by frederikhors** [reply]
@davidbarsky thank you very much. You're helping a passionate guy which is on the first day with Rust! Thank you very much!

**Reply by frederikhors** [reply]
I think I just found it:

![image](https://user-images.githubusercontent.com/41120635/182716119-9e2576aa-b1d3-48bc-920b-cf14bfdb5805.png)

**Reply by frederikhors** [reply]
Now only this is the remaining question, if you can help me better understand:

> A quick question: it's fine to use `tracing_subscriber::fmt().init()` directly or should I use `registry` with `tracing_subscriber::fmt::layer()`?

**Reply by davidbarsky** [reply]
yes, sorry! it's totally fine to use `tracing_subscriber::fmt().init()` instead of `fmt::Layer()` + the registry. `init()` does a few more things depending on Cargo's feature flags (sets up `log` integration and `EnvFilter` are the big ones), but it's possible to setup these with layers and the registry. the biggest advantage of using `fmt::Layer` and the Registry is that it makes it easy to add additional layers, such as `tracing-opentelemetry` or [`tracing-chrome`](https://crates.io/crates/tracing-chrome). In other words, it's a lower-level API, but a more flexible one.

**Reply by CEbbinghaus** [reply]
Sorry to necro this post but I am banging my head against the wall to get this working.
If I use `tracing-subscriber` without any dependencies it fails to find the json method (ofc coz its a crate feature):
`tracing-subscriber = "0.3.18"`
```cargo
error[E0599]: no method named `json` found for struct `SubscriberBuilder` in the current scope
  --> src/log.rs:12:50
   |
12 |     let mut file_writer = tracing_subscriber::fmt().json().with_writer(file);
   |                                                     ^^^^ method not found in `SubscriberBuilder`
```

However enabling the feature causes completely different (already functioning) methods to break:
`tracing-subscriber = { version = "0.3.18", features = ["default", "json"] }` (the default is there for sanity)
```
error[E0599]: no method named `set_ansi` found for struct `SubscriberBuilder` in the current scope
  --> src/log.rs:14:14
   |
14 |     file_writer.set_ansi(false);
   |                 ^^^^^^^^ method not found in `SubscriberBuilder<JsonFields, Format<Json>, LevelFilter, File>`

error[E0599]: no method named `with_filter` found for struct `SubscriberBuilder` in the current scope
  --> src/log.rs:18:16
   |
18 |             file_writer.with_filter(tracing_subscriber::filter::LevelFilter::from_level(CONFIG.log_level))
   |                         ^^^^^^^^^^^ help: there is a method with a similar name: `with_file`
```

This also happens when I go overboard and enable all features at once:
```
tracing-subscriber = { version = "0.3.18", features = [
    "default",
    "alloc",
    "std",
    "env-filter",
    "fmt",
    "ansi",
    "registry",
    "json",
] }
```

I'm at my witts end. No idea why enabling Json breaks everything else

**Reply by mladedav** [reply]
@CEbbinghaus could you share the code? Are you absolutely sure you're using the same versions?

**Reply by CEbbinghaus** [reply]
I am using the same version, Although I since managed to figure it out with the help of a Rust community member. What I had not considered is that By changing the code to use the `json()` method I had to change the namespace & therefore the type of the instance. As such they were valid errors that just appeared invalid as the type they were on was different. That was 110% my fault for not recognizing the failures as genuine failures due to bad code and not as failures of the library.

Here is the change in question although it was after I had figured it out & therefore is very small:
https://github.com/CEbbinghaus/MicroSDeck/pull/11/commits/7a2fdf317879ebe1e645c465d20ac083dc418b10
